pymc_marketing/customer_choice/nested_logit.py

```diff
@@ -0,0 +1,783 @@
+        result: dict[str, int | float | dict] = {
+            "alphas_": self.model_config["alphas_"].to_dict(),
+            "likelihood": self.model_config["likelihood"].to_dict(),
+            "betas": self.model_config["betas"].to_dict(),
+            "betas_fixed": self.model_config["betas_fixed_"].to_dict(),
+        }
+
+        return result
+
+    def parse_formula(self, df, formula, depvar):
+        """Parse the three-part structure of a formula specification.
+
+        Splits the formula into target, alternative-specific covariates, and
+        fixed covariates. Ensures that the target variable appears in the
+        dependent variable column and that all specified covariates exist
+        in the input dataframe.
+        """
+        target, covariates = formula.split("~")
+        target = target.strip()
+
+        alt_covariates, fixed_covariates = covariates.split("|")
+        alt_covariates = alt_covariates.strip()
+        fixed_covariates = fixed_covariates.strip()
+
+        if target not in df[depvar].unique():
+            raise ValueError(
+                f"Target '{target}' not found in dependent variable '{depvar}'."
+            )
+
+        for f in fixed_covariates.split("+"):
+            if f.strip() and f.strip() not in df.columns:
+                raise ValueError(
+                    f"Fixed covariate '{f.strip()}' not found in dataframe columns."
+                )
+
+        for a in alt_covariates.split("+"):
+            if a.strip() and a.strip() not in df.columns:
+                raise ValueError(
+                    f"Alternative covariate '{a.strip()}' not found in dataframe columns."
+                )
+
+        return target, alt_covariates, fixed_covariates
+
+    def prepare_X_matrix(self, df, utility_formulas, depvar):
+        """Prepare the X matrix for the utility equations.
+
+        The X matrix is a tensor with dimensions:
+        (N observations) x (Alternatives) x (Covariates).
+        Assumes that the utility formulas specify an equal number of covariates
+        per alternative; these can be zero-valued if an alternative lacks a
+        specific attribute.
+
+        Utility formulas should express the relationship between the choice
+        outcome (dependent variable) and the attributes of each alternative
+        that incentivize that choice. The left-hand side (LHS) of each formula
+        must correspond to a value of the dependent variable, while the
+        right-hand side (RHS) should define an additive combination of the
+        available covariates.
+
+        We also allow the incorporation of fixed covariates that do not vary
+        across alternatives. For these, an alternative-specific parameter is
+        used to allow the contribution to utility to vary by alternative.
+        """
+        n_obs = len(df)
+        n_alts = len(utility_formulas)
+        n_covariates = len(utility_formulas[0].split("|")[0].split("+"))
```

Comment:
maybe this can be encoded in a function = len(_get_covariates(...)). This way we can unit-test `_get_covariates`?